pandas/core/groupby/generic.py

```diff
@@ -856,23 +856,28 @@ def _get_index():
             return index
 
         if isinstance(values[0], dict):
-            # GH #823
+            # GH #823 #24880
             index = _get_index()
-            result = DataFrame(values, index=index).stack()
+            result = self._reindex_output(DataFrame(values, index=index))
```

Comment:
Is there a test case that hits this? Transforms by definition broadcast to match the shape of the caller so didn't think anything should change within this function

Reply:
The change is in `_wrap_applied_output`. I did not modify `_wrap_transformed_output`, only `_wrap_aggregated_output` and `_wrap_applied_output`. Tests that hit this are `test_groupby_observed` and `test_groupby_observed_apply_lambda_returns_dict`.